### PR TITLE
Improve JDTLS defaults and improve documentation

### PR DESCRIPTION
**Pull request by eruizc-dev**
After the discussion in #1067 this is what came up:

 - Improve default root_dir: 
    - Some people complaint that projects without package managers didn't work, so I added Makefile and .git
    - The multi-file project root was unnecessary, it also included `.kts` files that are for Kottlin projects
 - Improve cmd autodetect:
    - Removed `cmd_env` property that did nothing
    - Made `JAVA_HOME` and `WORKSPACE` env variables optional
    - Unified `JAR` and `JDTLS_CONFIG` into `JDTLS_HOME`
    - Preserved `JAR` and `JDTLS_CONFIG` with a message to avoid breaking existing configurations
    - Added `.exe` for windows support
 - Update documentation
    - Add links to point in the right direction
    - Mention jdtls-launcher as unofficial installer/launcher
    - Mention nvim-jdtls as complete client implementation
    - Provide actual configuration examples

### CODE REVIEW

**Comment by eruizc-dev** [comment]
I've been using it all day with no issues whatsoever (in linux), here's my config:

Neovim config:
https://github.com/eruizc-dev/eruizc-dev/blob/jdtls-pr-test/dotfiles/xdg_config_home/nvim/lua/eruizc/lsp.lua#L95

Environment:
https://github.com/eruizc-dev/eruizc-dev/blob/jdtls-pr-test/dotfiles/.zshrc#L79-L81

File: lua/lspconfig/jdtls.lua [hunk 1]
```diff
@@ -129,26 +140,50 @@ configs[server_name] = {
...
+IMPORTANT: If you want all the features jdtls has to offer, [nvim-jdtls](https://github.com/mfussenegger/nvim-jdtls)
+is highly recommended. If all you need is diagnostics, completion, imports, gotos and formatting and some code actions
+you can keep reading here.
+
+For manual installation you can download precompiled binaries from the
+[official downloads site](http://download.eclipse.org/jdtls/snapshots/?d)
 
-Due to the nature of java, the settings for eclipse jdtls cannot be automatically
-inferred. Please set the following environmental variables to match your installation. You can set these locally for your project with the help of [direnv](https://github.com/direnv/direnv). Note version numbers will change depending on your project's version of java, your version of eclipse, and in the case of JDTLS_CONFIG, your OS.
+Due to the nature of java settings cannot be inferred, please set the following
```
Comment (mjlbach) [inline lua/lspconfig/jdtls.lua:154]: You removed a comma that should probably not have been removed

File: lua/lspconfig/jdtls.lua [hunk 2]
```diff
@@ -129,26 +140,50 @@ configs[server_name] = {
...
+# Optional:
+export JAVA_HOME=/path/to/java_home # In case you don't have java in path or want to use a version in particular
+export WORKSPACE=/path/to/workspace # Defaults to $HOME/workspace
+```
+```lua
+  -- init.lua
+  require'lspconfig'.jdtls.setup{}
 ```
+
+For automatic instllation you can use the following unofficial installers/launchers under your own risk:
```
Comment (mjlbach) [inline lua/lspconfig/jdtls.lua:172]: instllation

File: lua/lspconfig/jdtls.lua [hunk 3]
```diff
@@ -129,26 +140,50 @@ configs[server_name] = {
...
+  - [jdtls-launcher](https://github.com/eruizc-dev/jdtls-launcher) (Includes lombok support by default)
```
Comment (mjlbach) [inline lua/lspconfig/jdtls.lua:173]: Until `curl -s https://raw.githubusercontent.com/eruizc-dev/jdtls-launcher/master/install.sh | sudo bash` is removed from that launcher, I can't recommend it.

File: lua/lspconfig/jdtls.lua [hunk 4]
```diff
@@ -71,44 +82,44 @@ end
...
-    root_dir = function(fname)
-      for _, patterns in ipairs(root_files) do
-        local root = util.root_pattern(unpack(patterns))(fname)
-        if root then
-          return root
-        end
-      end
-      return vim.fn.getcwd()
-    end,
+    root_dir = util.root_pattern({
```
Comment (mjlbach) [inline lua/lspconfig/jdtls.lua:102]: People asked for the mult-root fallback, so I'm concerned about changing it.

File: lua/lspconfig/jdtls.lua [hunk 5]
```diff
@@ -1,61 +1,72 @@
...
---- Callback function for the `language/status` notification.
---
--- The server sends a non-standard notification when the status of the language
--- server changes. This can be used to display progress as the server is
--- starting up.
+  if vim.fn.getenv('JDTLS_HOME') == vim.NIL then
+    error('JDTLS_HOME env variable not set')
+  end
+
+  if vim.fn.has('unix') ~= 0 then
```
Comment (mjlbach) [inline lua/lspconfig/jdtls.lua:48]: I would use luv's os_uname here instead

File: lua/lspconfig/jdtls.lua [hunk 6]
```diff
@@ -1,61 +1,72 @@
...
+    print('Using old $JDTLS_CONFIG environment variable for JDTLS, refer to documentation and update to JDTLS_HOME')
+    return vim.fn.getenv('JDTLS_CONFIG')
+  end
 
---- Callback function for the `language/status` notification.
---
--- The server sends a non-standard notification when the status of the language
--- server changes. This can be used to display progress as the server is
--- starting up.
+  if vim.fn.getenv('JDTLS_HOME') == vim.NIL then
```
Comment (mjlbach) [inline lua/lspconfig/jdtls.lua:44]: os.getenv

File: lua/lspconfig/jdtls.lua [hunk 7]
```diff
@@ -1,61 +1,72 @@
...
+  if vim.fn.getenv('JAR') ~= vim.NIL then
+    print('Using old $JAR environment variable for JDTLS, refer to documentation and update to JDTLS_HOME')
+    return vim.fn.getenv('JAR')
+  end
+
+  if vim.fn.getenv('JDTLS_HOME') == vim.NIL then
+    error('JDTLS_HOME env variable not set')
+  end
+
+  return vim.fn.expand('$JDTLS_HOME/plugins/org.eclipse.equinox.launcher_*.jar')
```
Comment (mjlbach) [inline lua/lspconfig/jdtls.lua:34]: os.getenv + concatenation

Comment (mjlbach) [reply lua/lspconfig/jdtls.lua:34]: Uh, I get you are doing a glob... maybe it's fine

File: lua/lspconfig/jdtls.lua (hunk 1 above)
Comment (eruizc-dev) [reply lua/lspconfig/jdtls.lua]: Fixed

File: lua/lspconfig/jdtls.lua (hunk 2 above)
Comment (eruizc-dev) [reply lua/lspconfig/jdtls.lua]: Fixed

File: lua/lspconfig/jdtls.lua [hunk 8]
```diff
@@ -1,33 +1,35 @@
+local os = require 'os'
```
Comment (mjlbach) [inline lua/lspconfig/jdtls.lua:1]: this is unecessary

File: lua/lspconfig/jdtls.lua [hunk 9]
```diff
@@ -1,33 +1,35 @@
...
 local configs = require 'lspconfig/configs'
 local util = require 'lspconfig/util'
 local handlers = require 'vim.lsp.handlers'
 
 local server_name = 'jdtls'
 
 local function get_java_executable()
-  local executable = vim.fn.getenv('JAVA_HOME') ~= vim.NIL
-    and vim.fn.expand('$JAVA_HOME/bin/java')
+  local executable = os.getenv('JAVA_HOME') ~= vim.NIL
```
Comment (mjlbach) [inline lua/lspconfig/jdtls.lua:10]: I would just allocate a local variable for home here. Also you can test against nil, not vim.nil, and simplify this to if home then...

File: lua/lspconfig/jdtls.lua (hunk 4 above)
Comment (eruizc-dev) [reply lua/lspconfig/jdtls.lua]: I didn't saw the getcwd and added it back. I don't see why the multi root would be necessary, are there any issues related to that?

Comment (mjlbach) [reply lua/lspconfig/jdtls.lua]: https://github.com/neovim/nvim-lspconfig/issues/455

Comment (mjlbach) [reply lua/lspconfig/jdtls.lua]: Also I got asked about this on matrix IIRC and dm'd a few times

File: lua/lspconfig/jdtls.lua (hunk 3 above)
Comment (eruizc-dev) [reply lua/lspconfig/jdtls.lua:178]: I've been working on this and now sudo is no longer necessary.
https://github.com/eruizc-dev/jdtls-launcher#linux--macos--wsl

File: lua/lspconfig/jdtls.lua (hunk 5 above)
Comment (eruizc-dev) [reply lua/lspconfig/jdtls.lua:50]: I couldn't find this function anywhere, where should I look for this?

File: lua/lspconfig/jdtls.lua (hunk 8 above)
Comment (eruizc-dev) [reply lua/lspconfig/jdtls.lua:1]: Thank you for pointing out, didn't know that!

File: lua/lspconfig/jdtls.lua (hunk 5 above)
Comment (mjlbach) [reply lua/lspconfig/jdtls.lua:50]: With neovim, you have access to all of luv's functions (bindings to libuv) under vim.loop

https://github.com/luvit/luv/blob/master/docs.md#uvos_uname

vim.loop == uv for those examples

Comment (mjlbach) [reply lua/lspconfig/jdtls.lua:50]: You can grep the repo for how we match against windows/other systems

File: lua/lspconfig/jdtls.lua (hunk 7 above)
Comment (eruizc-dev) [reply lua/lspconfig/jdtls.lua:36]: I couldn't find another way of achieving this without iterate files and doing tons of checks. The filename goes like this: `org.eclipse.equinox.launcher_1.6.200.v20210416-2027.jar`.

Comment (mjlbach) [reply lua/lspconfig/jdtls.lua:36]: it's probably fine, you could list the directory with luv, iterate and :match on the predicates org.eclipse.equinox.launcher to avoid going through vimscript, but it's ok for now to use vim.fn.expand

**Comment by mjlbach** [comment]
A couple high level comments for cleanup:

* store envvars in local variables and reuse those, it makes the code much cleaner
* vim.NIL is not nil in lua land
* I'm still not sure about all of these deprecation notices, we have the breaking changes issue now

**Comment by eruizc-dev** [comment]
I addressed all things you mentioned, let me know what you think!

**Comment by eruizc-dev** [comment]
Sry I had to force push, I forgot something as unstaged in the last commit. It's fixed now and ready for review!

**Comment by eruizc-dev** [comment]
I fixed the lint ci but commit-lint will fail again. Should I rename all commits, squash them or is it okay to leave like this? You can always squash at the moment of merging.

**Comment by mjlbach** [comment]
I do always squash if a PR doesn't follow our commit guidelines, I added the CI recently with the help of dundargoc to encourage users to squash their own commits so I don't have to guesstimate the correct squashed message :)

**Comment by eruizc-dev** [comment]
There ya go, commits squashed!

File: lua/lspconfig/jdtls.lua [hunk 10]
```diff
@@ -1,61 +1,53 @@
...
-    'settings.gradle', -- Gradle
-    'settings.gradle.kts', -- Gradle
-  },
-  -- Multi-module projects
-  { 'build.gradle', 'build.gradle.kts' },
-}
+local function get_java_executable()
+  local executable = env.JAVA_HOME and env.JAVA_HOME .. '/bin/java' or 'java'
+
+  return sysname:match 'Windows' and executable .. '.exe' or executable
```
Comment (mfussenegger) [inline lua/lspconfig/jdtls.lua:18]: Will the path concattenation above actually work under windows? Before this used `util.path.join` 
(Same also happens in other places)

Comment (eruizc-dev) [reply lua/lspconfig/jdtls.lua:18]: I'm currently doing some tests in Windows, will confirm once it works

**Comment by eruizc-dev** [comment]
Regarding Windows, I managed to get it to execute with the correct params but I get an exit code 13... This is what LspInfo tells me, there are no logs whatsoever.

According to my google search, this relates to Java environment not being properly set up. I've never worked with Java in Windows and there's something I probably did wrong.

image

**Comment by mfussenegger** [comment]
@mjlbach as far as I can tell this now looks okay.

**Comment by bndbsh** [comment]
@eruizc-dev this breaks when jdtls is installed by a package manager. jdtls wants write access to the config directory. Previously, it was possible to set the config directory separately from path to the jar. Now it's mandatory to install jdtls in your home directory in order to use this. This should at least be noted in the instructions, though ideally we'd revert to being able to set the paths separately.

**Comment by eruizc-dev** [comment]
@bndbsh I wasn't aware of any package managers that included jdtls. And I didn't know they wouldn't give you access to config directories. I'll fix that, just tell me what package manager you use!

**Comment by bndbsh** [comment]
@eruizc-dev package managers generally install to a system-wide directory. I installed [this](https://aur.archlinux.org/packages/jdtls/).

I think it might be enough to just allow overriding the config directory with `JDTLS_CONFIG`, and if it's not present keep the current behavior.

**Comment by mjlbach** [comment]
@bndbsh Please open another issue to track this
